File: Strings/392. Is Subsequence/392. Is Subsequence.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isSubsequence = function (s, t) {
  if (!s) return true;
  if (s.length > t.length) return false;

  let s_idx = 0;
  let t_idx = 0;

  while (t_idx < t.length) {
    if (s[s_idx] === t[t_idx]) {
      s_idx++;
      if (s_idx === s.length) return true;
    }
    t_idx++;
  }

  return false;
};

// Time: O(n)
// Space: O(1)

File: Strings/392. Is Subsequence/392. Is Subsequence.py
class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        if not s:
            return True
        
        if len(s) > len(t):
            return False

        s_idx = t_idx = 0
        while t_idx < len(t):
            if s[s_idx] == t[t_idx]:
                s_idx += 1
                if s_idx == len(s):
                    return True
            t_idx += 1
        
        return False
                
# Time: O(n)
# Space: O(1)